File: pm_trader/portfolio.py
"""Portfolio orchestrator — turn the pool scanner + single-pool bot into a system.

The validated edge is capacity-constrained: each two-sided ``min_size`` quote earns
a few $/day but locks only ~$50-100, so the strategy is "spread a small book across
many uncorrelated SAFE mid-tail pools."  This module does exactly that:

  scan → select fundable SAFE pools within a capital budget → one dry-run maker bot
  per pool → aggregate expected reward, committed capital, and per-poll plans.

Dry-run only (no real orders); composes ``scanner`` + ``live`` + ``reward_math``.
"""

from __future__ import annotations

import re

from pm_trader.maker_live import LiveMakerBot
from pm_trader.orderbook import committed_capital, maker_quote_score, maker_reward_share
from pm_trader.rewards import RewardsClient, scan

DEFAULT_CAPITAL = 1000.0

# Generic words that carry no correlation signal (so two markets sharing only
# these are NOT treated as the same event).
_STOPWORDS = frozenset(
    "will the be in on at a an of to and or by win wins won most next be the for "
    "vs end up down above below before after into out as is are reach hit with "
    "who what when which than then over under between".split()
)


def _significant_tokens(question: str) -> set[str]:
    """Distinctive lowercase tokens of a question (entities/places), minus
    stopwords, pure numbers, and very short tokens — used to detect correlation."""
    toks = re.findall(r"[a-z0-9]+", (question or "").lower())
    return {t for t in toks if len(t) > 2 and not t.isdigit() and t not in _STOPWORDS}


# Curated topic/entity clusters: markets matching the same cluster move together
# (same FOMC, same country's politics, same race) even when they share no surface
# words — e.g. "Grindeanu next PM" and "Bolojan out" are both the Romanian PM.
# Surface-token overlap can't catch those, so we cluster on known correlated themes.
# Extend as new correlated themes appear in the reward universe.
_CLUSTER_RULES: tuple[tuple[str, tuple[str, ...]], ...] = (
    ("us-fed", ("fed", "fomc", "federal reserve")),
    ("romania", ("romania", "grindeanu", "bolojan", "ciolacu")),
    ("colombia", ("colombia", "petro")),
    ("israel", ("israel", "netanyahu", "knesset", "eizenkot", "bennett")),
    ("france", ("france", "french", "bardella", "macron", "philippe", "melenchon")),
    ("ecb", ("ecb", "european central bank", "lagarde")),
    ("boe", ("bank of england", "boe")),
)


def _cluster_key(question: str) -> str | None:
    """Canonical correlation cluster for a question, or None if it matches no
    known correlated theme.  Catches same-event/same-country pairs that share no
    surface words.  Heuristic + curated — it cannot catch every entity link
    (e.g. an unlisted politician's name), so it complements, not replaces, the
    token-overlap check."""
    q = (question or "").lower()
    for name, keywords in _CLUSTER_RULES:
        if any(kw in q for kw in keywords):
            return name
    return None


def _risk_adjusted_score(p: dict, risk_tolerance_days: float,
                         chop_aversion: float = 1.0) -> float:
    """Reward per day, discounted by BOTH jump-tail AND steady-chop exposure.

    ``reward_per_day`` is the steady income. Two risk dimensions shrink it:
      - jump tail: ``days_wiped`` = days of income one historical MAX jump erases.
        FULLY weighted — a discrete gap fills before any cancel, so reaction speed
        cannot make jumpy pools safe.
      - steady chop: ``daily_vol_c`` = std of typical daily mid moves (cents),
        normalised by the band (``max_spread_c``). Continuous chop IS cancellable,
        so ``chop_aversion`` (0..1) scales this penalty down for a fast canceller
        (1.0 = a slow poller's full aversion; 0.5 = half). Ranking thus favours
        high-yield AND jump-safe pools, with chop-tolerance tuned to our latency.
    """
    reward = p.get("reward_per_day")
    if reward is None:
        reward = p.get("share", 0.0) * p.get("daily", 0.0)
    days_wiped = p.get("days_wiped")
    if days_wiped is None:
        days_wiped = 9_999.0
    jump_disc = 1.0 + days_wiped / risk_tolerance_days
    vol = p.get("daily_vol_c") or 0.0           # 0/None (no history) -> no chop penalty
    band = p.get("max_spread_c") or 0.0
    chop_disc = 1.0 + chop_aversion * (vol / band if band > 0 else 0.0)
    return reward / (jump_disc * chop_disc)


def _deploy_size(p: dict, half_spread_c: float, *, target_cap: float,
                 share_cap: float, loss_budget: float) -> float:
    """Capital-aware order size for a pool, >= min_size, capped by THREE limits so
    sizing up never quietly blows risk:
      - capital: don't lock more than ``target_cap`` (the per-pool capital slice).
      - reward-share: keep our est share <= ``share_cap`` — reward is ~linear in size
        only while our share is small; past that we'd dominate the pool, the reward
        saturates, and we become the adverse-selection target.
      - jump risk: a single worst-case historical jump (``max_jump_c``) on this size
        must not exceed ``loss_budget`` — so a gap we CAN'T cancel stays survivable.
    """
    min_size = p["min_size"]
    per_share = committed_capital(min_size, half_spread_c) / min_size  # capital lock / share
    cap_capital = (target_cap / per_share) if per_share > 0 else min_size
    w = maker_quote_score(1.0, half_spread_c, p.get("max_spread_c") or 0.0)  # in-band weight
    comp = p.get("min_side_score") or 0.0                                    # competitors' Qmin
    cap_share = (share_cap / (1.0 - share_cap) * comp / w) if (w > 0 and 0 < share_cap < 1) else cap_capital
    mj = (p.get("max_jump_c") or 0.0) / 100.0
    cap_risk = (loss_budget / mj) if mj > 0 else cap_capital
    return max(min_size, min(cap_capital, cap_share, cap_risk))


def select_pools(
    scan_report: dict,
    *,
    capital: float = DEFAULT_CAPITAL,
    require_safe: bool = True,
    half_spread_ticks: int = 1,
    risk_tolerance_days: float = 7.0,
    chop_aversion: float = 1.0,
    max_token_overlap: int = 1,
    size_share_cap: float = 0.33,
    loss_budget: float = 0.0,
    quality_floor_frac: float = 0.10,
    max_pool_frac: float = 0.25,
    cooldown: set | None = None,
) -> list[dict]:
    """Pick a fundable, diversified book by RISK-ADJUSTED yield within a budget.

    Selection criteria, in order:
      0. Cooldown: skip any pool (by condition_id or token) in ``cooldown`` — a
         market that just had a catalyst jump is no longer the quiet SAFE pool we
         classified, so don't re-enter it until it has settled.
      1. Eligibility: SAFE + non-empty-band (unless ``require_safe=False``).
      2. Rank by RISK-ADJUSTED yield (``reward_per_day`` discounted by jump-tail
         exposure) — not cheapest-first, not gross-yield-only.
      3. Correlation filter: skip a candidate that is in the same curated topic
         cluster (Fed, a country's politics, …) as an already-selected pool, OR
         shares more than ``max_token_overlap`` distinctive tokens with one
         (same race / event / entity ⇒ correlated ⇒ no real diversification).
      4. Fund greedily down the ranked list until the capital budget runs out.

    Sizing is always yield-weighted: capital flows to quality (size ∝ score),
    capped per pool by ``max_pool_frac`` plus the share/jump limits in
    ``_deploy_size``; a tight budget / zero ``loss_budget`` floors it at each
    pool's ``min_size``. Pool COUNT is governed by the dynamic quality floor +
    budget — there is no hard pool-count cap.
    """
    cd = cooldown or set()
    cands = [
        p for p in scan_report.get("pools", [])
        if p.get("condition_id") not in cd and p.get("token") not in cd
        and (not require_safe or (p.get("jump_verdict") == "SAFE" and not p.get("empty_band")))
    ]
    scored = [(p, _risk_adjusted_score(p, risk_tolerance_days, chop_aversion)) for p in cands]
    scored.sort(key=lambda ps: -ps[1])
    # DYNAMIC quality floor (not a hard pool count): keep funding pools while they're
    # worth at least `quality_floor_frac` of the BEST pool's risk-adjusted score. A
    # clear quality gap -> fund only the few good ones; a flat field of good pools ->
    # fund many. This drops the junk tail so capital isn't tied up in weak pools.
    best = scored[0][1] if scored else 0.0
    cutoff = quality_floor_frac * best if quality_floor_frac > 0 else 0.0
    elig = [(p, s) for p, s in scored if s >= cutoff]
    total_score = sum(s for _, s in elig) or 1.0

    selected: list[dict] = []
    chosen_tokens: list[set[str]] = []
    chosen_clusters: set[str] = set()
    spent = 0.0
    for p, score in elig:
        half_spread_c = p["tick"] * 100.0 * half_spread_ticks
        # YIELD-WEIGHTED allocation: capital flows to quality (size ∝ score), with a
        # per-pool diversification cap (max_pool_frac) so no pool dominates, plus the
        # share/jump caps in _deploy_size. Risk budget scales with the allocation.
        # A tight budget / zero loss_budget floors size at the pool's min_size.
        weight = score / total_score
        target = min(max_pool_frac * capital, capital * weight)
        rp = loss_budget * weight if loss_budget > 0 else 0.0
        size = _deploy_size(p, half_spread_c, target_cap=target,
                            share_cap=size_share_cap, loss_budget=rp)
        share = maker_reward_share(size, half_spread_c, p["max_spread_c"],
                                   p.get("min_side_score") or 0.0)
        cap = committed_capital(size, half_spread_c)
        if cap <= 0 or spent + cap > capital:
            continue
        cluster = _cluster_key(p["question"])
        if cluster is not None and cluster in chosen_clusters:
            continue  # same curated topic/country cluster ⇒ correlated
        toks = _significant_tokens(p["question"])
        if toks and any(len(toks & ct) > max_token_overlap for ct in chosen_tokens):
            continue  # surface-token correlation
        selected.append({
            "question": p["question"],
            "condition_id": p["condition_id"],
            "token": p["token"],
            "daily": p["daily"],
            "share": round(share, 4),
            "min_size": p["min_size"],
            "size": round(size, 2),
            "tick": p["tick"],
            "max_spread_c": p["max_spread_c"],
            "half_spread_c": half_spread_c,
            "committed_capital": round(cap, 2),
            "est_daily_reward": round(share * p["daily"], 4),
            "risk_adj_score": round(score, 4),
        })
        spent += cap
        chosen_tokens.append(toks)
        if cluster is not None:
            chosen_clusters.add(cluster)
    return selected


class MakerPortfolio:
    """A portfolio of two-sided maker quotes across selected pools.

    ``dry_run=True`` (default) plans orders without sending. For LIVE, pass
    ``dry_run=False`` and a real ``submitter`` (e.g. ``build_clob_signer()``);
    each bot then places/cancels real orders through it.
    """

    def __init__(self, selected: list[dict], *, dry_run: bool = True,
                 submitter=None) -> None:
        self.meta: dict[str, dict] = {}
        self.bots: dict[str, LiveMakerBot] = {}
        for s in selected:
            self.meta[s["token"]] = s
            self.bots[s["token"]] = LiveMakerBot(
                token_id=s["token"], max_spread_c=s["max_spread_c"],
                min_size=s.get("size") or s["min_size"], tick=s["tick"],
                half_spread_c=s["half_spread_c"], dry_run=dry_run,
                submitter=submitter, external_fills=not dry_run,
            )

    def committed_capital(self) -> float:
        return round(
            sum(committed_capital(b.size, b.half_spread_c) for b in self.bots.values()), 2
        )

    def est_daily_reward(self) -> float:
        return round(sum(m["est_daily_reward"] for m in self.meta.values()), 4)

    def plan_all(self, market_data: dict) -> list[dict]:
        """Run one quoting step per pool. ``market_data``: {token: (OrderBook, mid)}."""
        plans = []
        for token, bot in self.bots.items():
            md = market_data.get(token)
            if md is None:
                continue
            book, mid = md
            plans.append({"question": self.meta[token]["question"], **bot.step(book, mid)})
        return plans

    def summary(self) -> dict:
        cap = self.committed_capital()
        daily = self.est_daily_reward()
        return {
            "pools": len(self.bots),
            "committed_capital": cap,
            "free_implied": None,
            "est_daily_reward": daily,
            "est_annualized_pct": round(daily * 365 / cap * 100, 1) if cap > 0 else 0.0,
        }


def fetch_market_data(client: RewardsClient, selected: list[dict]) -> dict:
    """Pull a live (OrderBook, mid) for each selected pool's token, for plan_all.

    Returns {token: (OrderBook, mid)}; pools whose book is empty/one-sided or
    errors are skipped (so a single bad market doesn't sink the portfolio).
    """
    from pm_trader.models import OrderBook, OrderBookLevel

    out: dict = {}
    for s in selected:
        token = s["token"]
        try:
            raw = client.book(token)
        except Exception:
            continue
        bids = [OrderBookLevel(price=float(b["price"]), size=float(b["size"]))
                for b in (raw.get("bids") or []) if "price" in b and "size" in b]
        asks = [OrderBookLevel(price=float(a["price"]), size=float(a["size"]))
                for a in (raw.get("asks") or []) if "price" in a and "size" in a]
        if not bids or not asks:
            continue
        mid = (max(b.price for b in bids) + min(a.price for a in asks)) / 2.0
        out[token] = (OrderBook(bids=bids, asks=asks), mid)
    return out


def build_portfolio(
    client: RewardsClient,
    *,
    capital: float = DEFAULT_CAPITAL,
    min_daily: float = 80.0,
    top: int = 40,
) -> dict:
    """Scan live, select a fundable SAFE book, and return its plan + summary (dry-run)."""
    report = scan(client, min_daily=min_daily, top=top, with_jump_risk=True)
    selected = select_pools(report, capital=capital)
    portfolio = MakerPortfolio(selected, dry_run=True)
    market_data = fetch_market_data(client, selected)
    plans = portfolio.plan_all(market_data)
    return {
        "capital_budget": capital,
        "selected": selected,
        "summary": portfolio.summary(),
        "planned": len(plans),
        "plans": plans,
    }


def run(**kwargs: object) -> dict:
    """Convenience wrapper: build a client, build the portfolio, close the client."""
    client = RewardsClient()
    try:
        return build_portfolio(client, **kwargs)  # type: ignore[arg-type]
    finally:
        client.close()

File: cpp/src/pmm/portfolio.cpp
// src/pmm/portfolio.cpp — 选池配资编排实现 (port of pm_trader/portfolio.py 的纯选择部分)
#include "pmm/portfolio.hpp"

#include <algorithm>
#include <set>
#include <string>
#include <utility>
#include <vector>

#include "pmm/orderbook.hpp"
#include "pmm/round.hpp"

namespace pmm::portfolio {

namespace ob = pmm::orderbook;

namespace {

// 无相关性信号的通用词 (两市场只共享这些 ⇒ 不算同一事件)。
const std::set<std::string>& stopwords() {
    static const std::set<std::string> kSet = {
        "will", "the",  "be",   "in",     "on",    "at",    "a",     "an",    "of",    "to",
        "and",  "or",   "by",   "win",    "wins",  "won",   "most",  "next",  "for",   "vs",
        "end",  "up",   "down", "above",  "below", "before","after", "into",  "out",   "as",
        "is",   "are",  "reach","hit",    "with",  "who",   "what",  "when",  "which", "than",
        "then", "over", "under","between"};
    return kSet;
}

// 精选话题/实体簇: 命中同簇的市场一起动 (同 FOMC / 同国政治 / 同场比赛)。
const std::vector<std::pair<std::string, std::vector<std::string>>>& cluster_rules() {
    static const std::vector<std::pair<std::string, std::vector<std::string>>> kRules = {
        {"us-fed", {"fed", "fomc", "federal reserve"}},
        {"romania", {"romania", "grindeanu", "bolojan", "ciolacu"}},
        {"colombia", {"colombia", "petro"}},
        {"israel", {"israel", "netanyahu", "knesset", "eizenkot", "bennett"}},
        {"france", {"france", "french", "bardella", "macron", "philippe", "melenchon"}},
        {"ecb", {"ecb", "european central bank", "lagarde"}},
        {"boe", {"bank of england", "boe"}},
    };
    return kRules;
}

std::string to_lower(std::string s) {
    for (char& c : s) c = static_cast<char>((c >= 'A' && c <= 'Z') ? c + 32 : c);
    return s;
}

bool all_digits(const std::string& s) {
    if (s.empty()) return false;
    for (char c : s) {
        if (c < '0' || c > '9') return false;
    }
    return true;
}

std::size_t intersection_size(const std::set<std::string>& a, const std::set<std::string>& b) {
    std::size_t n = 0;
    for (const auto& x : a) {
        if (b.count(x) != 0) ++n;
    }
    return n;
}

// _deploy_size: 资本感知的下单 size (>= min_size), 受 资本/份额/跳跃 三限。
double deploy_size(const rewards::PoolReport& p, double half_spread_c, double target_cap,
                   double share_cap, double loss_budget) {
    const double min_size = p.min_size;
    const double per_share = ob::committed_capital(min_size, half_spread_c) / min_size;
    const double cap_capital = per_share > 0.0 ? (target_cap / per_share) : min_size;
    const double w = ob::maker_quote_score(1.0, half_spread_c, p.max_spread_c);
    const double comp = p.min_side_score;  // 竞争者 Qmin (PoolReport 总有值; None→0 等价)
    const double cap_share =
        (w > 0.0 && share_cap > 0.0 && share_cap < 1.0)
            ? (share_cap / (1.0 - share_cap) * comp / w)
            : cap_capital;
    const double mj = p.max_jump_c.value_or(0.0) / 100.0;
    const double cap_risk = mj > 0.0 ? (loss_budget / mj) : cap_capital;
    return std::max(min_size, std::min({cap_capital, cap_share, cap_risk}));
}

}  // namespace

std::set<std::string> significant_tokens(const std::string& question) {
    const std::string q = to_lower(question);
    std::set<std::string> out;
    std::string cur;
    auto flush = [&] {
        if (cur.size() > 2 && !all_digits(cur) && stopwords().count(cur) == 0) out.insert(cur);
        cur.clear();
    };
    for (char c : q) {
        if ((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')) {
            cur.push_back(c);
        } else {
            flush();
        }
    }
    flush();
    return out;
}

std::string cluster_key(const std::string& question) {
    const std::string q = to_lower(question);
    for (const auto& [name, keywords] : cluster_rules()) {
        for (const auto& kw : keywords) {
            if (q.find(kw) != std::string::npos) return name;
        }
    }
    return {};
}

double risk_adjusted_score(const rewards::PoolReport& p, double risk_tolerance_days,
                           double chop_aversion, double compet_aversion) {
    const double reward = p.reward_per_day;  // PoolReport 总有 reward_per_day (= share*daily, 已 round)
    const double days_wiped = p.days_wiped.value_or(9999.0);  // None(无历史) → 9999
    const double jump_disc = 1.0 + days_wiped / risk_tolerance_days;
    const double vol = p.daily_vol_c.value_or(0.0);  // 0/None → 无 chop 惩罚
    const double band = p.max_spread_c;
    const double chop_disc = 1.0 + chop_aversion * (band > 0.0 ? (vol / band) : 0.0);
    // 竞争度惩罚: PM 官方 market_competitiveness (实测 0.46–4.2, 越高越挤 = 我们份额被稀释越狠)。
    // 拥挤池降权 → 偏好清静池 (实测好选 0x551db8 comp0.47, 烂选 0xdf2020 comp4.2 小池)。<0=未知, 不罚。
    const double comp_disc =
        (compet_aversion > 0.0 && p.competitiveness >= 0.0) ? (1.0 + compet_aversion * p.competitiveness) : 1.0;
    return reward / (jump_disc * chop_disc * comp_disc);
}

std::vector<SelectedPool> select_pools(const rewards::ScanResult& scan_report,
                                       const SelectParams& params) {
    const auto& cd = params.cooldown;

    // 1. 资格过滤 (cooldown + SAFE/non-empty-band)
    std::vector<const rewards::PoolReport*> cands;
    for (const auto& p : scan_report.pools) {
        if (cd.count(p.condition_id) != 0 || cd.count(p.token) != 0) continue;
        if (params.require_safe && !(p.jump_verdict == "SAFE" && !p.empty_band)) continue;
        cands.push_back(&p);
    }

    // 2. 按风险调整收益排序 (稳定, 降序)
    std::vector<std::pair<const rewards::PoolReport*, double>> scored;
    scored.reserve(cands.size());
    for (const auto* p : cands) {
        scored.push_back({p, risk_adjusted_score(*p, params.risk_tolerance_days, params.chop_aversion,
                                                  params.compet_aversion)});
    }
    std::stable_sort(scored.begin(), scored.end(),
                     [](const auto& a, const auto& b) { return a.second > b.second; });

    // 3. 动态质量地板 (不是硬池数): 保留 >= quality_floor_frac * 最优分 的池
    const double best = scored.empty() ? 0.0 : scored.front().second;
    const double cutoff = params.quality_floor_frac > 0.0 ? params.quality_floor_frac * best : 0.0;
    std::vector<std::pair<const rewards::PoolReport*, double>> elig;
    for (const auto& ps : scored) {
        if (ps.second >= cutoff) elig.push_back(ps);
    }
    double total_score = 0.0;
    for (const auto& ps : elig) total_score += ps.second;
    if (total_score == 0.0) total_score = 1.0;

    // 4*. 注水配资 (waterfill): 先按分数+去相关选池、每池下 min_size, 再把剩余资本按"边际 κ×奖励/美元"
    //     注水。奖励份额对 size 是凹的 (share=ours/(ours+E) 饱和) → 边际递减 → 注水均衡边际 = 最大化组合奖励。
    //     取代"capital ∝ score"(无视凹性, 会过度配资到已饱和的高分池)。+7.9% (量化备忘), 现按校准净值。
    if (params.waterfill) {
        const double infl = 1.0 / std::max(params.reward_calib, 1e-6);  // 竞争充气 = 校准真实份额
        struct WF {
            const rewards::PoolReport* p;
            double hs, per_share, size, cap, max_size, score;
        };
        std::vector<WF> wf;
        std::vector<std::set<std::string>> chosen_tokens;
        std::set<std::string> chosen_clusters;
        double spent = 0.0;
        for (const auto& [p, score] : elig) {  // 4a. 选池 (分数序, 去相关) + min_size 起步
            const double hs = p->tick * 100.0 * params.half_spread_ticks;
            const double per_share = ob::committed_capital(p->min_size, hs) / p->min_size;
            const double min_cap = ob::committed_capital(p->min_size, hs);
            if (min_cap <= 0.0 || spent + min_cap > params.capital) continue;
            const std::string cluster = cluster_key(p->question);
            if (!cluster.empty() && chosen_clusters.count(cluster) != 0) continue;
            const std::set<std::string> toks = significant_tokens(p->question);
            bool corr = false;
            for (const auto& ct : chosen_tokens)
                if (static_cast<long>(intersection_size(toks, ct)) > params.max_token_overlap) {
                    corr = true;
                    break;
                }
            if (corr) continue;
            const double rp = params.loss_budget > 0.0 ? params.loss_budget / static_cast<double>(elig.size()) : 1e18;
            const double max_size = std::max(deploy_size(*p, hs, params.capital, params.size_share_cap, rp), p->min_size);
            wf.push_back({p, hs, per_share, p->min_size, min_cap, max_size, score});
            spent += min_cap;
            chosen_tokens.push_back(toks);
            if (!cluster.empty()) chosen_clusters.insert(cluster);
        }
        const double chunk = std::max(1.0, params.capital * 0.005);  // 4b. 注水剩余资本
        auto marg = [&](const WF& a) -> double {  // 边际 κ×奖励/美元 (再投 ~chunk 的奖励增量/资本增量)
            if (a.size >= a.max_size - 1e-9 || a.per_share <= 1e-9) return 0.0;
            const double dsize = std::min(chunk / a.per_share, a.max_size - a.size);
            const double s0 = ob::maker_reward_share(a.size, a.hs, a.p->max_spread_c, a.p->min_side_score * infl);
            const double s1 = ob::maker_reward_share(a.size + dsize, a.hs, a.p->max_spread_c, a.p->min_side_score * infl);
            return (s1 - s0) * a.p->daily / (dsize * a.per_share);
        };
        while (params.capital - spent > chunk && !wf.empty()) {
            int bi = -1;
            double bm = 0.0;
            for (std::size_t i = 0; i < wf.size(); ++i) {
                const double m = marg(wf[i]);
                if (m > bm) {
                    bm = m;
                    bi = static_cast<int>(i);
                }
            }
            if (bi < 0) break;  // 全部到顶
            WF& a = wf[static_cast<std::size_t>(bi)];
            const double dsize = std::min(chunk / a.per_share, a.max_size - a.size);
            const double dcap = dsize * a.per_share;
            if (dcap <= 1e-9) break;
            a.size += dsize;
            a.cap += dcap;
            spent += dcap;
        }
        std::vector<SelectedPool> out;  // 4c. 构建
        for (const auto& a : wf) {
            const double share = ob::maker_reward_share(a.size, a.hs, a.p->max_spread_c, a.p->min_side_score * infl);
            SelectedPool sp;
            sp.question = a.p->question;
            sp.condition_id = a.p->condition_id;
            sp.token = a.p->token;
            sp.daily = a.p->daily;
            sp.share = round_to(share, 4);
            sp.min_size = a.p->min_size;
            sp.size = round_to(a.size, 2);
            sp.tick = a.p->tick;
            sp.max_spread_c = a.p->max_spread_c;
            sp.half_spread_c = a.hs;
            sp.committed_capital = round_to(a.cap, 2);
            sp.est_daily_reward = round_to(share * a.p->daily, 4);  // 校准份额 × daily
            sp.risk_adj_score = round_to(a.score, 4);
            out.push_back(std::move(sp));
        }
        return out;
    }

    // 4. 贪心配资 (相关性/簇去重 + 预算)
    std::vector<SelectedPool> selected;
    std::vector<std::set<std::string>> chosen_tokens;
    std::set<std::string> chosen_clusters;
    double spent = 0.0;

    for (const auto& [p, score] : elig) {
        const double half_spread_c = p->tick * 100.0 * params.half_spread_ticks;
        // 始终按质量加权配资 (capital ∝ score, max_pool_frac 封顶 + deploy_size 的份额/跳变上限);
        // 预算紧/loss_budget=0 时自动退化到 min_size。池数由质量地板+预算治理, 无硬池数上限。
        const double weight = score / total_score;
        const double target = std::min(params.max_pool_frac * params.capital, params.capital * weight);
        const double rp = params.loss_budget > 0.0 ? params.loss_budget * weight : 0.0;
        const double size = deploy_size(*p, half_spread_c, target, params.size_share_cap, rp);
        const double share = ob::maker_reward_share(size, half_spread_c, p->max_spread_c, p->min_side_score);
        const double cap = ob::committed_capital(size, half_spread_c);
        if (cap <= 0.0 || spent + cap > params.capital) continue;

        const std::string cluster = cluster_key(p->question);
        if (!cluster.empty() && chosen_clusters.count(cluster) != 0) continue;  // 同簇 ⇒ 相关

        const std::set<std::string> toks = significant_tokens(p->question);
        if (!toks.empty()) {
            bool correlated = false;
            for (const auto& ct : chosen_tokens) {
                // 有符号比较 (对齐 Python len(...) > max_token_overlap); 负的 max_token_overlap 才不会回绕。
                if (static_cast<long>(intersection_size(toks, ct)) > params.max_token_overlap) {
                    correlated = true;
                    break;
                }
            }
            if (correlated) continue;
        }

        SelectedPool sp;
        sp.question = p->question;
        sp.condition_id = p->condition_id;
        sp.token = p->token;
        sp.daily = p->daily;
        sp.share = round_to(share, 4);
        sp.min_size = p->min_size;
        sp.size = round_to(size, 2);
        sp.tick = p->tick;
        sp.max_spread_c = p->max_spread_c;
        sp.half_spread_c = half_spread_c;
        sp.committed_capital = round_to(cap, 2);
        sp.est_daily_reward = round_to(share * p->daily, 4);
        sp.risk_adj_score = round_to(score, 4);
        selected.push_back(std::move(sp));

        spent += cap;
        chosen_tokens.push_back(toks);
        if (!cluster.empty()) chosen_clusters.insert(cluster);
    }
    return selected;
}

}  // namespace pmm::portfolio
